New snapshot save function and multi-gpu tests

kan-bayashi opened the pull request:
Save trainer / model / optimizer states into one file in pytorch.

- Added `torch_snapshot` to save trainer / model / optimizer states into one file
- Added `torch_resume` to resume training from new snapshot
- Unified `torch_save` and `torch_load` functions in ASR and TTS
- Added multi-gpu tests for pytorch
- Added save / load test for pytorch
- Added more docstrings

sw005320 commented on src/asr/asr_utils.py:
```diff
@@ -235,6 +204,87 @@ def _plot_and_save_attention(self, att_w, filename):
+    s.save(trainer)
+    if hasattr(trainer.updater.model, "model"):
+        # (for TTS)
+        if hasattr(trainer.updater.model.model, "module"):
```
Why TTS has a bit redundant model structure (`model.model`)?
Is there a way to make it same as ASR?

kan-bayashi replied:
If we assume that same options are used when resuming, it will be more simple.

But if you want to keep same format as model.loss.best, current format is needed.

kan-bayashi commented:
@sw005320 Made more simple. which is better?
Previous: may work if we change options when resuming, but complex
Now: not work if we change options but simple

sw005320 commented:
My preference is previous one.
The code becomes complex in `asr_utils.py`, but this is OK.

kan-bayashi commented:
Reverted.

kan-bayashi commented:
Updated.
- Replaced loading part in `asr_pytorch.recog` with `torch_load`
- Update `torch_load`
    - Both model file and snapshot file to be used in decoding phase
- Added some cleaning part (remove loaded object explicitly)

kan-bayashi commented:
Added more test codes.

kan-bayashi commented:
@sw005320 I want to merge. Is it OK?
